iOS support

erikchi opened the pull request:
This PR is for adding the iOS support and forcing iOS to use interpreter mode for now.

erikchi committed:
gomobile bind for iOS

erikchi committed:
fix: changed one API to  , and if is_ios then create a specific config to force use interpreter mode (avoid JIT for now in iOS)

erikchi commented on dialer.go:
```diff
@@ -27,15 +34,36 @@ type Dialer struct {
 }
 
-func NewDialer() *Dialer {
+func NewDialer(is_ios bool) *Dialer {
```
This can later be changed to a switch for users to choose which mode they want to use (mainly for Android rn, iOS need some more dig for the compiler mode to work).

gaukas commented on Makefile:
```diff
@@ -1,16 +1,16 @@
 IMPORT_PATH=github.com/gaukas/watermob
 
-BUILD_IOS="cd $(BUILDDIR) && $(GOBIND) -a -ldflags $(LDFLAGS) -target=$(IOS_TARGET) -o $(IOS_ARTIFACT) $(IMPORT_PATH)"
+BUILD_IOS="$(GOBIND) -target=$(IOS_TARGET) -o $(IOS_ARTIFACT)"
```
Is this change at all necessary? (and by what means)

erikchi replied:
Yeah, the previous command doesn't work for building iOS, so I changed to this. And I didn't add the `-ldflags` for now.

gaukas replied:
If at all possible we should at least debug this to learn why it will not work. 

The original command `cd` into the build output dir at `$(BUILDDIR)` and builds the source in `$(IMPORT_PATH)` which should be resolved by `gomobile` automatically after identifying the package name. 

Meanwhile at this point we might want to make build command for Android and iOS consistent. So I will push a commit to update the Android command into the same style.

erikchi replied:
Thanks for the alignment! Yeah, I'll try to see what happened with the previous command async so not block this PR to be merged.

gaukas replied:
Thanks!

gaukas commented on dialer.go:
```diff
@@ -27,15 +34,36 @@ type Dialer struct {
 		unprotectedDial: net.Dial,
 	}
+
+	if is_ios {
+		// Create a ConfigJSON object with only ForceInterpreter set
+		config := ConfigJSONiOS{
+			Runtime: struct {
+				ForceInterpreter bool `json:"force_interpreter,omitempty"`
+			}{
+				ForceInterpreter: true, // Setting this to true or false as needed
+			},
+		}
+
+		// Marshal the config object to JSON
+		jsonData, err := json.Marshal(config)
+		if err != nil {
+			panic(fmt.Sprintf("Error marshaling JSON:", err))
+		}
+
+		dialer.configJSON = jsonData
+	}
```
This change looks redundant. Should it be possible for us to programmatically detect iOS and switch `ForceInterpreter` to `true` without using the JSON (JSON was meant to be reserved for out-sourced customization).

erikchi replied:
For now I inserted a bool to distinguish iOS and Android in this API (`NewDialer`) from the Go part. And we can definitely eliminating the usage of JSON to set this where I can move this bool to `dialWATER` where we can first check if we have customized JSON / Proto to set things, otherwise if it's from iOS, then we create with the interpreter mode there.

gaukas replied:
I was thinking about something else. Like, we can base on build tag to set a flag to true for iOS and directly call `*(water.Config).RuntimeConfig().Interpreter()`

gaukas replied:
I will try to push a change reflecting what I described.

erikchi replied:
Thanks for the changes, I think it looks great! With testing, `runtime.GOOS == "ios"` worked perfectly.

gaukas commented:
I'm trying to get a push on this PR and hopefully get this merged ASAP, since this PR seems to be a blocker to our upcoming add-on of benchmark suites.

gaukas committed:
update: detect iOS with runtime.GOOS

Instead of setting configJSON when creating the Dialer base on a hardcoded flag, a more elegant (and less conflicting) way might be detecting the presence of iOS using runtime.GOOS and directly enable the interpreter mode on the finalized Config.

Signed-off-by: Gaukas Wang <i@gaukas.wang>

gaukas committed:
update: consistent Makefile command style

Signed-off-by: Gaukas Wang <i@gaukas.wang>

gaukas commented:
@erikziyunchi Can you check again and make sure my change: 
- [x] does not break it for iOS build with the newly added `LDFLAGS`
- [x] has equivalently same behavior for the compiled artifact in terms of enabling interpreter mode on iOS (without setting the boolean flag)

gaukas commented on dialer.go:
```diff
@@ -30,12 +37,14 @@ type Dialer struct {
 	water.SetGlobalCompilationCache(compilationCache)
 
-	return &Dialer{
+	var dialer = Dialer{
```
I feel this change can be reverted as well.

erikchi committed:
revert extra changes of the return part of NewDialer
